### PR TITLE
sync: expose strong and weak counts of mpsc sender handles

### LINKED ISSUE #5880: Add `mpsc::Sender` strong and weak counts, similar to `Arc` counts
**Is your feature request related to a problem? Please describe.**
The tokio actor pattern described in @Darksonn's [blog post](https://ryhl.io/blog/actors-with-tokio/) is very useful and pervasive. With #4023 we also have [WeakSender](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.WeakSender.html), automatically making [Sender](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html) a "strong" Sender. But if we have strong and weak senders, shouldn't we also expose the relevant counts as well?

**Describe the solution you'd like**
Just how Arc does with [strong](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.strong_count) and [weak](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.weak_count) counts, tokio's `mpsc::Sender` should expose similar counts.

**Describe alternatives you've considered**
We could probably wrap these types using newtype pattern, implement our custom clone and achieve the same result, if we desperately needed that. But it would be nice if official API offered these counts

**Additional context**
We have extended Alice's initial [actor pattern](https://ryhl.io/blog/actors-with-tokio/) with automatic actor termination when all relevant handles get dropped, graceful shutdown and on some actors even, we provide "weak" handles, implemented on top of [mpsc::WeakSender](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.WeakSender.html), offering similar functionality (doesn't prevent the actor to get dropped even if there are such "weak" handles around). It would be really helpful to have the counts as we could provide better API guarantees on graceful shutdown. Specifically, we wouldn't allow gracefully terminating an actor, even if a handle requests that, if other "strong" handles still exist.

### PR DESCRIPTION
**Pull request by maminrayej**
Resolves #5880.

### CODE REVIEW

File: tokio/src/sync/mpsc/unbounded.rs [hunk 1]
```diff
@@ -588,13 +598,29 @@ impl<T> Clone for WeakUnboundedSender<T> {
     }
 }
 
+impl<T> Drop for WeakUnboundedSender<T> {
+    fn drop(&mut self) {
+        self.chan.decrement_weak_count();
+    }
+}
```
Comment (Darksonn) [inline tokio/src/sync/mpsc/unbounded.rs:605]: Do you not also need an increment in clone?

Please add a test that would have caught this.

Comment (maminrayej) [reply tokio/src/sync/mpsc/unbounded.rs:607]: You're right.

I modified the test and can confirm that it fails before the fix.

File: tokio/tests/sync_mpsc_weak.rs [hunk 2]
```diff
@@ -511,3 +512,161 @@ fn test_tx_count_weak_unbounded_sender() {
 
     assert!(tx_weak.upgrade().is_none() && tx_weak2.upgrade().is_none());
 }
+
+#[tokio::test]
```
Comment (mox692) [inline tokio/tests/sync_mpsc_weak.rs:516]: Regarding tests, I personally like the one feature per one test style. In this case, we could split into these tests:

* strong_count is incremented when the Sender is cloned
* weak_count is incremented when the Sender is downgraded
* strong_count is decremented when the Sender is dropped
* weak_count is decremented when the WeakSender is dropped
* mixed cases not covered by the above, if needed

**Review by mox692 — APPROVED** [review]
I left just one comment, but I don't think it block the merge.

File: tokio/tests/sync_mpsc_weak.rs (hunk 2 above)
Comment (maminrayej) [reply tokio/tests/sync_mpsc_weak.rs:515]: That's a good point

**Review by Darksonn — APPROVED** [review]

**Pull request merged**